etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,187 @@
+from typing import Set
+from typing import Union
+
+import scipy
+from scipy.sparse import lil_matrix
+
+from etna.core import BaseMixin
+
+
+class HierarchicalStructure(BaseMixin):
+    """Represents hierarchical structure for provided hierarchical tree."""
+
+    def __init__(self, level_structure: Dict[str, List[str]], level_names: Optional[List[str]] = None):
+        """Init HierarchicalStructure.
+
+        Parameters
+        ----------
+        level_structure:
+            Adjacency list describing the structure of the hierarchy tree (i.e. {"total":["X", "Y"], "X":["a", "b"], "Y":["c", "d"]}).
+        level_names:
+            Names of levels in the hierarchy in the order from top to bottom (i.e. ["total", "category", "product"]).
+            If None is passed, level names are generated automatically with structure "level_<level_index>".
+        """
+        self._hierarchy_root: Union[str, None] = None
+        self._hierarchy_interm_nodes: Set[str] = set()
+        self._hierarchy_leaves: Set[str] = set()
+
+        self.level_structure: Dict[str, List[str]] = level_structure
+
+        self._find_graph_structure(level_structure)
+        hierarchy_levels = self._find_hierarchy_levels(level_structure)
+        tree_depth = len(hierarchy_levels)
+
+        if level_names is None:
+            level_names = [f"level_{i}" for i in range(tree_depth)]
+
+        if len(level_names) != tree_depth:
+            raise ValueError("Length of `level_names` must be equal to hierarchy tree depth!")
+
+        self.level_names: List[str] = level_names
+        self._level_series: Dict[str, List[str]] = {level_names[i]: hierarchy_levels[i] for i in range(tree_depth)}
+        self._level_to_index: Dict[str, int] = {level_names[i]: i for i in range(tree_depth)}
+
+        self._sub_segment_size_map: Dict[str, int] = {k: len(v) for k, v in level_structure.items()}
+
+        self._segment_subtree_size_map: Dict[str, int] = self._get_subtree_sizes(hierarchy_levels)
+
+        self._segment_to_level: Dict[str, str] = {
+            segment: level for level in self._level_series for segment in self._level_series[level]
+        }
+
+    def _find_graph_structure(self, adj_list: Dict[str, List[str]]):
+        """Find hierarchy top level (root of tree)."""
+        children = set(chain(*adj_list.values()))
+        parents = set(adj_list.keys())
+
+        tree_roots = parents.difference(children)
+        if len(tree_roots) != 1:
+            raise ValueError("Invalid tree definition: unable to find root!")
+
+        self._hierarchy_interm_nodes = parents & children
+        self._hierarchy_leaves = children.difference(parents)
+
+        tree_root = tree_roots.pop()
+        self._hierarchy_root = tree_root
+
+    def _find_hierarchy_levels(self, hierarchy_structure: Dict[str, List[str]]):
+        """Traverse hierarchy tree to group segments into levels."""
+        nodes: Set[str] = self._hierarchy_interm_nodes | self._hierarchy_leaves
```

Comment:
Store the `_num_nodes` as the attribute, move its calculation to `_find_graph_structure `